### ISSUE TITLE
Train non-English TTS model with nemo_tts

### ISSUE THREAD

**Issue by dmitrytyrin** [NVIDIA/NeMo#323]
Dear team,
I try to train tacotron2 for non-English tts.
I use ~20 hours (90% for train, 10% for test) 22050Hz dataset.
My labels:
labels: [' ', '!', '"', "'", '(', ')', ',', '-', '.', ':', ';', '?',
        'А', 'Б', 'В', 'Г', 'Д', 'Е', 'Ё', 'Ж', 'З', 'И', 'Й', 'К', 'Л', 'М', 'Н', 'О',
        'П', 'Р', 'С', 'Т', 'У', 'Ф', 'Х', 'Ц', 'Ч', 'Ш', 'Щ', 'Ъ', 'Ы', 'Ь', 'Э', 'Ю', 'Я',
        'а', 'б', 'в', 'г', 'д', 'е', 'ё', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о',
        'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ъ', 'ы', 'ь', 'э', 'ю', 'я']

After 20k steps with default batch size and amp_opt_level O1 I get only noise in generated audio (using pretrained WaveGlowNM.pt). Texts from train/test jsons dumped with ensure_ascii=False parameter.

Here is information from TensorBoard:
<img width="365" alt="Screenshot 2020-02-04 at 10 20 15" src="https://user-images.githubusercontent.com/54152608/73722653-0a040e00-4738-11ea-8bb1-a51f787e880a.png">
<img width="343" alt="Screenshot 2020-02-04 at 10 21 48" src="https://user-images.githubusercontent.com/54152608/73722753-53545d80-4738-11ea-9c52-c3338cc298c6.png">
<img width="343" alt="Screenshot 2020-02-04 at 10 23 37" src="https://user-images.githubusercontent.com/54152608/73722809-6cf5a500-4738-11ea-9dd6-9d2d9243e3cf.png">

Generated spectrogram:
![spec_0](https://user-images.githubusercontent.com/54152608/73723011-01f89e00-4739-11ea-80aa-079018db360b.png)

What can be wrong with my experiment?

**Comment by blisc** [comment]
Can you try with O0?

**Comment by dmitrytyrin** [comment]
@blisc, I got the same result:
<img width="626" alt="Screenshot 2020-02-06 at 16 31 39" src="https://user-images.githubusercontent.com/54152608/73941530-80eb0380-48fe-11ea-9e4f-4a697b9d1fbc.png">

---------------------------------
Train mel predicted/target looks very similar:
<img width="372" alt="Screenshot 2020-02-06 at 16 36 27" src="https://user-images.githubusercontent.com/54152608/73941829-fd7de200-48fe-11ea-96ff-9ec248084132.png">

----------------------------------
But model shows bad result on test:
<img width="372" alt="Screenshot 2020-02-06 at 16 36 47" src="https://user-images.githubusercontent.com/54152608/73941906-1edece00-48ff-11ea-9e66-82c024193604.png">

**Comment by blisc** [comment]
Train predictions will almost always look good. If they don't, there is likely an issue in the TTS setup.
Looking at your mels, it seems like you have significant silences at the beginning and end of your audio clips, can you try turning on trim_silence in the data_loader?

**Comment by dmitrytyrin** [comment]
@blisc, thank you for your time!
Results are better (trim silences, batch size 30, 36k steps), but still only noise...
<img width="373" alt="Screenshot 2020-02-17 at 09 25 23" src="https://user-images.githubusercontent.com/54152608/74628748-7d117980-5167-11ea-8c11-e33237297107.png">

<img width="362" alt="Screenshot 2020-02-17 at 09 20 41" src="https://user-images.githubusercontent.com/54152608/74628612-286dfe80-5167-11ea-81e7-1b7d4fad37d6.png">

Test loss is very high:
<img width="347" alt="Screenshot 2020-02-17 at 09 21 02" src="https://user-images.githubusercontent.com/54152608/74628633-33289380-5167-11ea-9e09-bca22a5825a2.png">

I think I should try another dataset or train much longer if I decrease batch size.

**Comment by blisc** [comment]
Do you have a multi speaker dataset? Or a single speaker dataset? Tacotron 2 requires a single speaker dataset.
Other than that, I'm not too sure what else could be the issue.

**Comment by ican24** [comment]
Dear Team,

For non-English/Mandarin languages your TTS also NLP are not compatible!
I had tested them with  dozen datasets and never succeeded since the first day of your project. Meantime other/alternative frameworks are generating good voices with the same datasets.
So you need to investigate your source code and inform your users about  compatible issue with the custom (non-English/Mandarin) languages.
This way you may save a tons of time and valuable resources of your users.
If I am a wrong, please provide us any simple proof of compatibility. 

Thank you in advance 

@dmitrytyrin 
Уважаемый Дмитрий, напишите мне письмо на электронный адрес, в ответ я подскажу как генерировать голос быстро и без труда с помощью других программ, если Вы уверены, что у Вас корпус корректный.

**Comment by dmitrytyrin** [comment]
Dear @blisc,
let me share with you results from another experiment:
I took another dataset of single speaker (~30h of audio, https://ruslan-corpus.github.io), converted from 44kHz to 22050Hz and trim silences with sox. Then I removed sentences with characters not in my labels, striped first and last whitespaces, removed audios longer than >10 sec and check EOS.

My labels:
[' ', '!', ',', '-', '.', ':', ';', '?', 'А', 'Б', 'В', 'Г', 'Д', 'Е', 'Ё', 'Ж', 'З', 'И', 'Й', 'К', 'Л', 'М', 'Н', 'О', 'П', 'Р', 'С', 'Т', 'У', 'Ф', 'Х', 'Ц', 'Ч', 'Ш', 'Щ', 'Ъ', 'Ы', 'Ь', 'Э', 'Ю', 'Я', 'а', 'б', 'в', 'г', 'д', 'е', 'ё', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о', 'п', 'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ъ', 'ы', 'ь', 'э', 'ю', 'я']

After 26k steps of training with NeMo from master branch with default parameters, O0, batch size 48 and "trim silence" parameter "true" in config file I get this result:
<img width="666" alt="Screenshot 2020-03-02 at 09 12 57" src="https://user-images.githubusercontent.com/54152608/75650130-1fe8ee00-5c66-11ea-8f91-347982ff9b3a.png">

<img width="666" alt="Screenshot 2020-03-02 at 09 13 18" src="https://user-images.githubusercontent.com/54152608/75650144-28d9bf80-5c66-11ea-92c3-e965587e1287.png">

<img width="348" alt="Screenshot 2020-03-02 at 09 22 45" src="https://user-images.githubusercontent.com/54152608/75650594-6723ae80-5c67-11ea-90b6-b10ddb9799df.png">

<img width="348" alt="Screenshot 2020-03-02 at 09 22 31" src="https://user-images.githubusercontent.com/54152608/75650603-6c80f900-5c67-11ea-990c-783f638a9551.png">

And inference with these checkpoints gave me 5 identical mels for 5 different sentences:
![spec_0](https://user-images.githubusercontent.com/54152608/75650361-b3222380-5c66-11ea-8d11-672009d16c24.png)
(this is just gurgle noise)

Do you have any ideas what can be wrong?
Is it possible that there is a problem somewhere in code with non-ASCII characters (models tries to predict \u0412 \u043a\u0430)?

**Comment by blisc** [comment]
We have identified a bug in tacotron 2 for version 0.9.0. Can you clone the latest master and try running the experiments there keeping all of your latest data preprocessing?
You can install from master after cd into the NeMo directory and running either `pip install nemo_toolkit[all]` or `pip install -e ".[all]"`.
Note that the config files have slightly changed going from 0.9.0 to master.

**Comment by dmitrytyrin** [comment]
@blisc, I rebuild my environment with 0.10.0b3 nemo-toolkit.
Training process seems unstable: I tried 3 times with the same arguments but only in one training model converged.
<img width="619" alt="Screenshot 2020-03-07 at 08 07 10" src="https://user-images.githubusercontent.com/54152608/76137161-b93f4800-604a-11ea-8c4b-e5fba5f70b9f.png">

Once I get exploding loss (after ~10k loss started increase very much)

And once I got this:
<img width="355" alt="Screenshot 2020-03-07 at 08 10 06" src="https://user-images.githubusercontent.com/54152608/76137205-2521b080-604b-11ea-86c4-8faebf89c2bb.png">
<img width="617" alt="Screenshot 2020-03-07 at 08 11 07" src="https://user-images.githubusercontent.com/54152608/76137213-3d91cb00-604b-11ea-8f1a-ed1c3c04ce18.png">

1) During testing model I get warnings about max decoder steps. How can I avoid this?
2) I generated some audio with converged model and I noticed that model can't work with long words. How can I solve this?
3) Generated audios have noticeable background noise. How can I avoid it?

**Comment by Piteryo** [comment]
Сan confirm the issue. I tried to run Tacotron training on my own Russian corpus with default parameters - got only a noise. Then I tried to trim my audios harder, it helped a bit: after 15000 steps alignment started to resemble diagonal, but with a lot of noise around. After 90000 steps I still don't have good alignment, the synthesized audio only resembles target, but mixes the letters and provide bad pronunciation. At the same time, Tensorflow version of Tacotron (https://github.com/Rayhane-mamah/Tacotron-2) on the same dataset works well after 15000 steps.

**Comment by ican24** [comment]
Подобные упущения делают этот проект бесполезным для разработчиков, которые на самом деле нуждаются в синтезе речи для своих языков (не английски, китайский мандарин или японский!).
Все это напоминает работу пресловутых банков, которые предлагают кредиты для клиентов, которые абсолютно не нуждаются в них. 
Зачем пользователей английского или китайского загрузить свои GPU, если  в интернете полно обученные модели?
Я просто не понимаю русскоговорящих разработчиках этого проекта. Что им мешает выделить 30 минут, чтобы разобраться в элементарных багах проекта и беречь пользователей от безумной потери времени и нервов. 
Престижно конечно работать в крутых корпорациях, но что они скажут, если одноклассник или однокурсник спросят "как я могу пользоваться вашими продуктами?" во время дружеского общения.
Не знаю, надоели бесконечные неспешные попытки и неясность Tacotron2.
Создали игрушку для явного PR-а и заставляют пользователей бесполезно экспериментировать.
Я добивался результатов с помощью менее известной платформы синтеза речи без колдования и потери времени. Не идеально, как показывают в промо ролликах. зато очень быстро и надежно. 
Вот здесь видео с 4 искусственными голосами.
[https://www.youtube.com/watch?v=AiHr3h0QZC4](url)
Если интересно, расскажу как обучать Ваш датасет. Напишите  мне письмо на электронный адрес.
Недавно мне удалось привязать waveglow c text2mel, с которым я работаю, но пока значительно артефактов, которые надо устранять.

**Comment by blisc** [comment]
For training non-English TTS, I would recommend the following changes to the tacotron2 yaml file.

Please use the [tacotron2_preprocedefaults](https://github.com/NVIDIA/NeMo/tree/tacotron2_preprocedefaults) branch. Inside the tacotron2 yaml, please make the following changes:
- `parser: en` -> `parser: base`
- `fmax: &fmax 8000` -> `fmax: &fmax null`
- Under `AudioToMelSpectrogramPreprocessor`'s `init_params`, add `mel_norm: 0.`

If attention is still not being learned, try constraining the negative values of the mel spectrogram even further. Two values need to be changed:
- Under `AudioToMelSpectrogramPreprocessor`'s `init_params`, change `log_zero_guard_value: 1e-05` to a larger value such as `1e-2`
- Change the corresponding `pad_value: &pad_value -11.52` to `floor(ln(log_zero_guard_value))`, for `1e-2`, a  `pad_value: &pad_value -4.61` should work.

For reference, feel free to look at [tacotron2_dev_a.yaml](https://github.com/NVIDIA/NeMo/blob/tacotron2_preprocedefaults/examples/tts/configs/tacotron2_dev_a.yaml) or [tacotron2_dev_b.yaml](https://github.com/NVIDIA/NeMo/blob/tacotron2_preprocedefaults/examples/tts/configs/tacotron2_dev_b.yaml).

Please note that this branch is still a WIP and could have issues. Also, note that we do not provide WaveGlow checkpoints that match the suggested mel spectrogram changes.

**Comment by ican24** [comment]
Dear @blisc, 
Thank you for your efforts to fix this serious bag of tacotron2, which disallows to reach sufficient results for non-English languages!
